### PR TITLE
Load virtio_rng early in the game

### PR DESCRIPTION
**Pull request by haraldh**
If libgcrypt is initialized, it needs a entropy for its random pool.
Without virtio_rng loaded initialize_libgcrypt() takes forever and the
boot stalls for a long time.

### CODE REVIEW

**Comment by haraldh** [comment]
Oh, in FIPS mode we have an even bigger problem.

In RHEL, we have a [patch](https://git.centos.org/commitdiff/rpms!libgcrypt.git/f268f105748f455df30c79bb1b21bb94a6575f2d)  :

libgcrypt does this in the library constructor: 
```
+      if (fips_mode ())
+         _gcry_random_initialize (1);
```

So, if systemd is executed and linked against libgcrypt, we have a problem with low entropy 👎

**Comment by lsiudut** [comment]
And unfortunately it is when built with Lorax, so I guess in all of RH-related use cases. It kills VM provisioning completely.

**Comment by poettering** [comment]
> If libgcrypt is initialized, it needs a entropy for its random pool.
> Without virtio_rng loaded initialize_libgcrypt() takes forever and the
> boot stalls for a long time.

hmm, did it always do that? how come this only comes up now?

File: src/core/kmod-setup.c [hunk 1]
```diff
@@ -45,6 +46,12 @@ static void systemd_kmod_log(
 }
 #endif
 
+static bool vm_is_kvm_or_qemu(void) {
+        int r = detect_vm();
+
+        return (r == VIRTUALIZATION_KVM || r == VIRTUALIZATION_QEMU);
+}
```
Comment (poettering) [inline src/core/kmod-setup.c]: please use IN_SET() for cases like this:

```
return IN_SET(detect_vm(), VIRTUALIZATION_KVM, VIRTUALIZATION_QEMU);

File: src/core/kmod-setup.c [hunk 2]
```diff
@@ -68,6 +75,7 @@ int kmod_setup(void) {
                 /* netfilter is needed by networkd, nspawn among others, and cannot be autoloaded */
                 { "ip_tables", "/proc/net/ip_tables_names", false,  false,   NULL      },
 #endif
+                { "virtio_rng", NULL,                       false,  false,   vm_is_kvm_or_qemu },
```
Comment (poettering) [inline src/core/kmod-setup.c]: please add a comment why we do this

**Comment by poettering** [comment]
hmm, can we tune the detection of this functionality a bit? it appears the availability of this virtio iface is announced to the OS an this can be easily checked via /sys. Can we tweak the detction accordingly, and look for the actual virtio iface rather thank blanket qemu/kvm checks? That way we don#t misdetect qemu invocations without it, and other VM implementations can implement the same iface sooner or later and things will just work.

**Comment by haraldh** [comment]
@poettering for /sys checking, we would have to check every ```/sys/devices/pci*/*/modalias``` for 
```
pci:v00001AF4d00001005*
pci:v00001AF4d00001044*
```

**Comment by poettering** [comment]
modinfo virtio-rng suggests the the following match instead: "virtio:d00000004v*" btw, not pci matches?

but looping through those dirs should be pretty starightforward, no?

**Comment by haraldh** [comment]
https://github.com/systemd/systemd/pull/6710

is the alternative with sysfs parsing

**Comment by poettering** [comment]
Closing this one in favour of #6710

**Pull request closed without merging**